File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return findHeight(root)==-1?0:1;
    }

    int findHeight(TreeNode* root){
        if(root==NULL) return 0;
        int left=findHeight(root->left);
        int right=findHeight(root->right);
        if((left==-1)||(right==-1)||(abs(right-left)>1)){
            return -1;
        }
         
        return max(left,right)+1;

    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        return checkBalanced(root)!=-1;
    }

    private static int checkBalanced(TreeNode root){
        if(root==null){
            return 0;
        }
        int leftheight = checkBalanced(root.left);
        if(leftheight==-1) return -1;
        int rightheight = checkBalanced(root.right);
        if(rightheight==-1) return -1;
        if(Math.abs(leftheight-rightheight)>1){
            return -1;
        }

        return  Math.max(leftheight,rightheight)+1;
    }
}